### DISCUSSION TITLE
Single Provider for all Stores

### DISCUSSION THREAD

**Discussion by damianon** [pmndrs/zustand#1157] [Q&A]
We're creating multiple stores to keep each of them simple and on topic. This works out great throughout the code, but causes issues when testing. React testing library executes all tests in a single virtual DOM, so the state doesn't get reset in between tests. One solution we know from Jotai is to wrap the test render in a `<Provider>`. 

However zustand requires the store for the Provider which wouldn't work in our case because there is not only one store. Is there any way to wrap it without having to specify a single store and have the Context limited for all stores?

**Comment by dai-shi** [comment]
Not sure if this is what you want from the testing perspective, but here's an example:

```jsx
import { createContext, useContext } from 'react'
import { createStore, useStore } from 'zustand'

const fooStore = createStore(...)
const barStore = createStore(...)

const MyContext = createContext()

const Provider = ({ children }) => (
  <MyContext.Provider value={{ foo: fooStore, bar: barStore }}>
    {children}
  </MyContext.Provider>
)

const useMyStore = (storeName, selector) => {
  const stores = useContext(MyContext)
  const store = stores[storeName]
  return useStore(store, selector)
}
```

**Comment by damianon** [comment]
I don't think this is an optimal solution, because now I have to add additional logic to put all stores in the context and retrieve them for use. And for every new store, I then have to Ideally the provider shouldn't need specifying all the stores. In Jotai for example I can just wrap my tests in `<Provider>` and wherever I use atoms they are all encapsulated, without the test/Provider needing to know about the specifics.

But maybe that's just not possible with the way zustand works? In that case I'll go with Jotai, because the main difference (state outside of react) is not really a concern for my application

**Reply by dai-shi** [reply]
Yeah. Jotai works differently because Jotai's store is a WeakMap and atom values are initialized on demand.
